iac-extensions/kubernetes/src/main/java/org/sonar/iac/helm/tree/utils/GoTemplateAstHelper.java

```diff
@@ -19,97 +19,83 @@
+  static Stream<FieldNode> findValuePathNodes(GoTemplateTree tree, TextRange textRange) {
     var nodes = tree.root().children().stream()
-      .filter(hasOverlayingLocation(location))
+      .filter(node -> overlap(node.textRange(), textRange))
       .toList();
 
-    return nodesWithImmediateChildren(nodes).stream()
+    return nodesWithAllChildren(nodes).stream()
       .filter(FieldNode.class::isInstance)
       // Sometimes top-level nodes have a very broad range and some children can be actually outside the range.
-      .filter(hasOverlayingLocation(location))
+      .filter(node -> overlap(node.textRange(), textRange))
       .map(FieldNode.class::cast);
   }
 
-  public static List<ValuePath> findValuePaths(GoTemplateTree tree, TextRange range, String sourceText) {
-    var location = LocationImpl.fromTextRange(range, sourceText);
-
-    return findValuePathNodes(tree, location)
+  public static List<ValuePath> findValuePaths(GoTemplateTree tree, TextRange range) {
+    return findValuePathNodes(tree, range)
       .map(FieldNode::identifiers)
       .map(ValuePath::new)
       .toList();
   }
 
-  private static Stream<Node> findIncludeFunctionsFirstArg(GoTemplateTree tree, Location location) {
+  private static Stream<Node> findIncludeFunctionsFirstArg(GoTemplateTree tree, TextRange textRange) {
     return collectNodesByType(tree.root(), NodeType.NODE_COMMAND, CommandNode.class)
-      .filter(hasOverlayingLocation(location))
+      .filter(node -> overlap(node.textRange(), textRange))
       .filter(node -> isFunction(node, "include"))
       .map(cmd -> cmd.arguments().get(1));
   }
 
-  private static Stream<? extends Node> findToYamlNodes(GoTemplateTree tree, Location location) {
+  private static Stream<? extends Node> findToYamlNodes(GoTemplateTree tree, TextRange textRange) {
     return collectNodesByType(tree.root(), NodeType.NODE_COMMAND, CommandNode.class)
-      .filter(hasOverlayingLocation(location))
+      .filter(node -> overlap(node.textRange(), textRange))
       .filter(node -> isFunction(node, "toYaml"));
   }
 
-  private static Predicate<Node> hasOverlayingLocation(Location location) {
-    return (Node node) -> {
-      var position = location.position();
-      var length = location.length();
-      var nodePosition = node.location().position();
-      var nodeLength = node.location().length();
-      return !(nodePosition > position + length || nodePosition + nodeLength < position);
-    };
-  }
-
-  private static List<Node> nodesWithImmediateChildren(List<Node> nodes) {
-    List<Node> allNodes = new ArrayList<>(nodes);
-    for (var i = 0; i < allNodes.size(); i++) {
-      allNodes.addAll(allNodes.get(i).children());
-    }
-    return allNodes;
+  private static List<Tree> nodesWithAllChildren(List<Tree> nodes) {
+    return Stream.concat(
+      nodes.stream(),
+      nodes.stream().flatMap(node -> nodesWithAllChildren(node.children()).stream()))
+      .toList();
```

Comment:
All the usages of this method convert the `List` into a stream.
Also in the method itself there is a lot of back and forth from list to stream to list.
It would make sense to change the signature to return `Stream<Tree>`